pymc3/tests/test_distributions_random.py

```diff
@@ -525,6 +500,147 @@ def seeded_kumaraswamy_rng_fn(self):
+    expected_rv_op_params = {"mu": mu, "sigma": sigma, "lower": lower, "upper": upper}
+    reference_dist_params = {
+        "loc": mu,
+        "scale": sigma,
+        "a": (lower - mu) / sigma,
+        "b": (upper - mu) / sigma,
+    }
+    reference_dist = seeded_scipy_distribution_builder("truncnorm")
+    tests_to_run = [
+        "check_pymc_params_match_rv_op",
+        "check_pymc_draws_match_reference",
+        "check_rv_size",
+    ]
+
+
+class TestTruncatedNormalLowerTau(BaseTestDistribution):
+    pymc_dist = pm.TruncatedNormal
+    lower, upper, mu, tau = -2.0, np.inf, 0, 1.0
+    tau, sigma = get_tau_sigma(tau=tau, sigma=None)
+    pymc_dist_params = {"mu": mu, "tau": tau, "lower": lower}
+    expected_rv_op_params = {"mu": mu, "sigma": sigma, "lower": lower, "upper": upper}
+    reference_dist_params = {"loc": mu, "scale": sigma, "a": (lower - mu) / sigma, "b": upper}
+    reference_dist = seeded_scipy_distribution_builder("truncnorm")
+    tests_to_run = [
+        "check_pymc_params_match_rv_op",
+        "check_pymc_draws_match_reference",
+        "check_rv_size",
+    ]
+
+
+class TestTruncatedNormalUpperTau(BaseTestDistribution):
+    pymc_dist = pm.TruncatedNormal
+    lower, upper, mu, tau = -np.inf, 2.0, 0, 1.0
+    tau, sigma = get_tau_sigma(tau=tau, sigma=None)
+    pymc_dist_params = {"mu": mu, "tau": tau, "upper": upper}
+    expected_rv_op_params = {"mu": mu, "sigma": sigma, "lower": lower, "upper": upper}
+    reference_dist_params = {"loc": mu, "scale": sigma, "a": lower, "b": (upper - mu) / sigma}
+    reference_dist = seeded_scipy_distribution_builder("truncnorm")
+    tests_to_run = [
+        "check_pymc_params_match_rv_op",
+        "check_pymc_draws_match_reference",
+        "check_rv_size",
+    ]
+
+
+class BaseTestWald(BaseTestDistribution):
+    def wald_rng_fn(self, size, mu, lam, alpha, uniform_rng_fct, normal_rng_fct):
+        v = normal_rng_fct(size=size) ** 2
+        z = uniform_rng_fct(size=size)
+        value = (
+            mu
+            + (mu ** 2) * v / (2.0 * lam)
+            - mu / (2.0 * lam) * np.sqrt(4.0 * mu * lam * v + (mu * v) ** 2)
+        )
+        i = np.floor(z - mu / (mu + value)) * 2 + 1
+        value = (value ** -i) * (mu ** (i + 1))
+        return value + alpha
+
+    def seeded_wald_rng_fn(self):
+        uniform_rng_fct = functools.partial(
+            getattr(np.random.RandomState, "uniform"), self.get_random_state()
+        )
+        normal_rng_fct = functools.partial(
+            getattr(np.random.RandomState, "normal"), self.get_random_state()
+        )
+        return functools.partial(
+            self.wald_rng_fn, uniform_rng_fct=uniform_rng_fct, normal_rng_fct=normal_rng_fct
+        )
+
+    pymc_dist = pm.Wald
+    reference_dist = seeded_wald_rng_fn
+    tests_to_run = [
+        "check_pymc_params_match_rv_op",
+        "check_pymc_draws_match_reference",
+        "check_rv_size",
+    ]
+
+
+class TestWaldPureScipy(BaseTestDistribution):
```

Comment:
I would expect this to pass, but it doesn't. I wonder whether it fails because somehow the random sampler seeding is not returning the expected results when we have more than one distribution being sampled in `rng_fn`. In this case both `rng.normal(size=size)` and `rng.uniform(size=size)`

Reply:
If the implementation is not the same, it's not necessarily the case they would match. Maybe the order the rng methods are called in also matters? By the way, is there a reason why we can't simply delegate to the `scipy` method?

Reply:
> Maybe the order the rng methods are called in also matters

yeah I think that in that case the order of the rng mattered. I think that the Scipy implementation of the Wald distribution doesn't match the PyMC one though. The Numpy one seems to be right though, so I replace the custom sampling with the numpy one

https://numpy.org/doc/stable/reference/random/generated/numpy.random.wald.html